x/upgrade/abci.go

```diff
@@ -53,49 +53,83 @@ func BeginBlocker(k keeper.Keeper, ctx sdk.Context, _ abci.RequestBeginBlock) {
 		},
 	)
 
-	// To make sure clear upgrade is executed at the same block
+	// If the plan's block height has passed, then it must be the executed version
+	// All major and minor releases are REQUIRED to execute on the scheduled block height
 	if plan.ShouldExecute(ctx) {
 		// If skip upgrade has been set for current height, we clear the upgrade plan
 		if k.IsSkipHeight(ctx.BlockHeight()) {
-			skipUpgradeMsg := fmt.Sprintf("UPGRADE \"%s\" SKIPPED at %d: %s", plan.Name, plan.Height, plan.Info)
-			ctx.Logger().Info(skipUpgradeMsg)
-
-			// Clear the upgrade plan at current height
-			k.ClearUpgradePlan(ctx)
+			skipUpgrade(k, ctx, plan)
 			return
 		}
+		// If we don't have an upgrade handler for this upgrade name, then we need to shutdown
+		if !k.HasHandler(plan.Name) {
+			panicUpgradeNeeded(k, ctx, plan)
+		}
+		applyUpgrade(k, ctx, plan)
```

Comment:
based on the other discussion we were doing, if the upgrade plan gets cleared after applying upgrade, wouldnt this cause issue for generating apphash since it does affect the stored state?

Reply:
My initial thought was that the upgrade action was stateless since registered plans stay registered, but if we specifically clear stuff out upon running upgrades, this may cause issue.

Reply:
Got it - I updated with the new logic (see next commit)